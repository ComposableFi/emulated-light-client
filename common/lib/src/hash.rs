use base64::engine::general_purpose::STANDARD as BASE64_ENGINE;
use base64::Engine;
#[cfg(feature = "borsh")]
use borsh::maybestd::io;
use bytemuck::TransparentWrapper;

/// A cryptographic hash.
#[derive(
    Clone,
    Default,
    PartialEq,
    Eq,
    derive_more::AsRef,
    derive_more::AsMut,
    derive_more::From,
    derive_more::Into,
    bytemuck::TransparentWrapper,
)]
#[cfg_attr(
    feature = "borsh",
    derive(borsh::BorshSerialize, borsh::BorshDeserialize)
)]
#[as_ref(forward)]
#[into(owned, ref, ref_mut)]
#[repr(transparent)]
pub struct CryptoHash(pub [u8; CryptoHash::LENGTH]);

// TODO(mina86): Make the code generic such that CryptoHash::digest take generic
// argument for the hash to use.  This would then mean that Trie, Proof and
// other objects which need to calculate hashes would need to take that argument
// as well.
impl CryptoHash {
    /// Length in bytes of the cryptographic hash.
    pub const LENGTH: usize = 32;

    /// Default hash value (all zero bits).
    pub const DEFAULT: CryptoHash = CryptoHash([0; 32]);

    /// Returns a builder which can be used to construct cryptographic hash by
    /// digesting bytes.
    #[inline]
    pub fn builder() -> Builder {
        Builder::default()
    }

    /// Returns hash of given bytes.
    #[inline]
    pub fn digest(bytes: &[u8]) -> Self {
        Self::digestv(core::slice::from_ref(&bytes))
    }

    /// Returns hash of concatenation of given byte slices.
    ///
    /// This is morally equivalent to feeding all the slices into the builder
    /// one-by-one or concatenating them into a single buffer and hashing it in
    /// a single step.
    ///
    /// Depending on platform this call may be more efficient.  Most notably,
    /// Solana offers a vectorised syscall for calculating a SHA-2 256 digest
    /// and this method will pass the request directly to it.  Note that
    /// `solana` crate feature must be enabled for this Solana-specific
    /// optimisation to be implemented.
    #[inline]
    pub fn digestv(slices: &[&[u8]]) -> Self {
        Self(imp::digestv(slices))
    }

    /// Decodes a base64 string representation of the hash.
    pub fn from_base64(base64: &str) -> Option<Self> {
        // base64 API is kind of garbage.  In certain situations the output
        // buffer must be larger than the size of the decoded data or else
        // decoding will fail.
        let mut buf = [0; 34];
        match BASE64_ENGINE.decode_slice(base64.as_bytes(), &mut buf[..]) {
            Ok(CryptoHash::LENGTH) => {
                Some(Self(*stdx::split_array_ref::<32, 2, 34>(&buf).0))
            }
            _ => None,
        }
    }

    /// Creates a new hash with given number encoded in its first bytes.
    ///
    /// This is meant for tests which need to use arbitrary hash values.
    #[cfg(feature = "test_utils")]
    pub const fn test(num: usize) -> CryptoHash {
        let mut buf = [0; Self::LENGTH];
        let num = (num as u32).to_be_bytes();
        let mut idx = 0;
        while idx < buf.len() {
            buf[idx] = num[idx % num.len()];
            idx += 1;
        }
        Self(buf)
    }

    /// Returns a shared reference to the underlying bytes array.
    #[inline]
    pub fn as_array(&self) -> &[u8; Self::LENGTH] {
        &self.0
    }

    /// Returns a shared reference to the hash as slice of bytes.
    #[inline]
    pub fn as_slice(&self) -> &[u8] {
        &self.0[..]
    }

    /// Allocates vector with the contents of the hash.
    #[inline]
    pub fn to_vec(&self) -> alloc::vec::Vec<u8> {
        self.as_slice().to_vec()
    }
}

impl core::fmt::Display for CryptoHash {
    /// Encodes the hash as base64 and prints it as a string.
    fn fmt(&self, fmtr: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        const ENCODED_LENGTH: usize = (CryptoHash::LENGTH + 2) / 3 * 4;
        let mut buf = [0u8; ENCODED_LENGTH];
        let len =
            BASE64_ENGINE.encode_slice(self.as_slice(), &mut buf[..]).unwrap();
        // SAFETY: base64 fills the buffer with ASCII characters only.
        fmtr.write_str(unsafe { core::str::from_utf8_unchecked(&buf[..len]) })
    }
}

impl core::fmt::Debug for CryptoHash {
    /// Encodes the hash as base64 and prints it as a string.
    #[inline]
    fn fmt(&self, fmtr: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        core::fmt::Display::fmt(self, fmtr)
    }
}

impl<'a> From<&'a [u8; CryptoHash::LENGTH]> for CryptoHash {
    #[inline]
    fn from(hash: &'a [u8; CryptoHash::LENGTH]) -> Self {
        <&CryptoHash>::from(hash).clone()
    }
}

impl From<&'_ CryptoHash> for [u8; CryptoHash::LENGTH] {
    #[inline]
    fn from(hash: &'_ CryptoHash) -> Self {
        hash.0
    }
}

impl<'a> From<&'a [u8; CryptoHash::LENGTH]> for &'a CryptoHash {
    #[inline]
    fn from(hash: &'a [u8; CryptoHash::LENGTH]) -> Self {
        CryptoHash::wrap_ref(hash)
    }
}

impl<'a> From<&'a mut [u8; CryptoHash::LENGTH]> for &'a mut CryptoHash {
    #[inline]
    fn from(hash: &'a mut [u8; CryptoHash::LENGTH]) -> Self {
        CryptoHash::wrap_mut(hash)
    }
}

impl<'a> TryFrom<&'a [u8]> for &'a CryptoHash {
    type Error = core::array::TryFromSliceError;

    #[inline]
    fn try_from(hash: &'a [u8]) -> Result<Self, Self::Error> {
        <&[u8; CryptoHash::LENGTH]>::try_from(hash).map(Into::into)
    }
}

impl<'a> TryFrom<&'a mut [u8]> for &'a mut CryptoHash {
    type Error = core::array::TryFromSliceError;

    #[inline]
    fn try_from(hash: &'a mut [u8]) -> Result<Self, Self::Error> {
        <&mut [u8; CryptoHash::LENGTH]>::try_from(hash).map(Into::into)
    }
}

impl<'a> TryFrom<&'a [u8]> for CryptoHash {
    type Error = core::array::TryFromSliceError;

    #[inline]
    fn try_from(hash: &'a [u8]) -> Result<Self, Self::Error> {
        <&CryptoHash>::try_from(hash).cloned()
    }
}

#[cfg(not(any(feature = "solana-program", target_os = "solana")))]
mod imp {
    use sha2::Digest;

    pub(super) fn digestv(slices: &[&[u8]]) -> [u8; 32] {
        panic!("WTF");
        solana_program::msg!("WTFFFFFFFFFFFFFf");
        let mut state = sha2::Sha256::new();
        for bytes in slices {
            state.update(bytes);
        }
        state.finalize().into()
    }

    #[derive(Default)]
    pub(super) struct State(sha2::Sha256);

    impl State {
        #[inline]
        pub fn update(&mut self, bytes: &[u8]) {
            self.0.update(bytes)
        }

        #[inline]
        pub fn done(self) -> [u8; 32] {
            self.0.finalize().into()
        }
    }
}

#[cfg(any(feature = "solana-program", target_os = "solana"))]
mod imp {
    use alloc::vec::Vec;

    pub(super) fn digestv(slices: &[&[u8]]) -> [u8; 32] {
        // panic!("Good");
        // solana_program::msg!("GOOOOOOOOOOOOOD");
        solana_program::hash::hashv(slices).to_bytes()
    }

    #[derive(Default)]
    pub(super) struct State(Vec<u8>);

    impl State {
        #[inline]
        pub fn update(&mut self, bytes: &[u8]) {
            self.0.extend_from_slice(bytes)
        }

        #[inline]
        pub fn done(self) -> [u8; 32] {
            solana_program::hash::hashv(&[&self.0]).to_bytes()
        }
    }
}

/// Builder for the cryptographic hash.
///
/// The builder calculates the digest of bytes that it’s fed using the
/// [`Builder::update`] method.
///
/// This is useful if there are multiple discontiguous buffers that hold the
/// data to be hashed.  If all data is in a single contiguous buffer it’s more
/// convenient to use [`CryptoHash::digest`] instead.
#[derive(Default)]
pub struct Builder(imp::State);

impl Builder {
    /// Process data, updating the internal state of the digest.
    #[inline]
    pub fn update(&mut self, bytes: &[u8]) {
        self.0.update(bytes)
    }

    /// Finalises the digest and returns the cryptographic hash.
    #[inline]
    pub fn build(self) -> CryptoHash {
        CryptoHash(self.0.done())
    }
}

#[cfg(feature = "borsh")]
impl io::Write for Builder {
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
        self.update(buf);
        Ok(buf.len())
    }

    fn write_all(&mut self, buf: &[u8]) -> io::Result<()> {
        Ok(self.update(buf))
    }

    fn flush(&mut self) -> io::Result<()> {
        Ok(())
    }
}

#[test]
fn test_new_hash() {
    assert_eq!(CryptoHash::from([0; 32]), CryptoHash::default());

    // https://www.di-mgt.com.au/sha_testvectors.html
    let want = CryptoHash::from([
        0xe3, 0xb0, 0xc4, 0x42, 0x98, 0xfc, 0x1c, 0x14, 0x9a, 0xfb, 0xf4, 0xc8,
        0x99, 0x6f, 0xb9, 0x24, 0x27, 0xae, 0x41, 0xe4, 0x64, 0x9b, 0x93, 0x4c,
        0xa4, 0x95, 0x99, 0x1b, 0x78, 0x52, 0xb8, 0x55,
    ]);
    assert_eq!(want, CryptoHash::digest(b""));
    assert_eq!(want, CryptoHash::builder().build());
    let got = {
        let mut builder = CryptoHash::builder();
        builder.update(b"");
        builder.build()
    };
    assert_eq!(want, got);
    assert_eq!(want, CryptoHash::builder().build());

    let want = CryptoHash::from([
        0xba, 0x78, 0x16, 0xbf, 0x8f, 0x01, 0xcf, 0xea, 0x41, 0x41, 0x40, 0xde,
        0x5d, 0xae, 0x22, 0x23, 0xb0, 0x03, 0x61, 0xa3, 0x96, 0x17, 0x7a, 0x9c,
        0xb4, 0x10, 0xff, 0x61, 0xf2, 0x00, 0x15, 0xad,
    ]);
    assert_eq!(want, CryptoHash::digest(b"abc"));
    let got = {
        let mut builder = CryptoHash::builder();
        builder.update(b"a");
        builder.update(b"bc");
        builder.build()
    };
    assert_eq!(want, got);
}
