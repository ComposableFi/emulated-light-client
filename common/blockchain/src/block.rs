use lib::hash::CryptoHash;

type Result<T, E = borsh::maybestd::io::Error> = core::result::Result<T, E>;

/// A single block of the emulated blockchain.
///
/// Emulated block’s height and timestamp are taken directly from the host
/// chain.  Emulated blocks don’t have their own timestamps.
///
/// A block is uniquely identified by its hash which can be obtained via
/// [`Block::calc_hash`].
///
/// Each block belongs to an epoch (identifier by `epoch_id`) which describes
/// set of validators which can sign the block.  A new epoch is introduced by
/// setting `next_epoch` field; epoch becomes current one starting from the
/// following block.
#[derive(
    Clone, Debug, PartialEq, Eq, borsh::BorshSerialize, borsh::BorshDeserialize,
)]
pub struct Block<PK> {
    /// Version of the structure.  At the moment always zero byte.
    version: crate::common::VersionZero,

    /// Hash of the previous block.
    pub prev_block_hash: CryptoHash,
    /// Height of the emulated blockchain’s block.
    pub block_height: crate::BlockHeight,
    /// Height of the host blockchain’s block in which this block was created.
    pub host_height: crate::HostHeight,
    /// Timestamp of the host blockchani’s block in which this block was created.
    pub host_timestamp: u64,
    /// Hash of the root node of the state trie, i.e. the commitment
    /// of the state.
    pub state_root: CryptoHash,

    /// Hash of the block in which current epoch has been defined.
    ///
    /// Epoch determines validators set signing each block.  If epoch is about
    /// to change, the new epoch is defined in `next_epoch` field.  Then, the
    /// very next block will use current’s block hash as `epoch_id`.
    pub epoch_id: CryptoHash,

    /// If present, epoch *the next* block will belong to.
    pub next_epoch: Option<crate::Epoch<PK>>,
}

/// Error while generating new block.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum GenerateError {
    /// Host height went backwards.
    BadHostHeight,
    /// Host timestamp went backwards.
    BadHostTimestamp,
}

impl<PK: crate::PubKey> Block<PK> {
    /// Returns whether the block is a valid genesis block.
    pub fn is_genesis(&self) -> bool {
        self.prev_block_hash == CryptoHash::DEFAULT &&
            self.epoch_id == CryptoHash::DEFAULT
    }

    /// Calculates hash of the block.
    pub fn calc_hash(&self) -> CryptoHash {
        let mut builder = CryptoHash::builder();
        borsh::to_writer(&mut builder, self).unwrap();
        builder.build()
    }


    /// Signs the block.
    pub fn sign<S>(&self, signer: &S) -> PK::Signature
    where
        S: crate::validators::Signer<Signature = PK::Signature>,
    {
        signer.sign(self.calc_hash().as_slice())
    }

    /// Verifies signature for the block.
    #[inline]
    pub fn verify(&self, pk: &PK, signature: &PK::Signature) -> bool {
        pk.verify(self.calc_hash().as_slice(), signature)
    }

    /// Constructs next block.
    ///
    /// Returns a new block with `self` as the previous block.  Verifies that
    /// `host_height` and `host_timestamp` don’t go backwards but otherwise they
    /// can increase by any amount.  The new block will have `block_height`
    /// incremented by one.
    pub fn generate_next(
        &self,
        host_height: crate::HostHeight,
        host_timestamp: u64,
        state_root: CryptoHash,
        next_epoch: Option<crate::Epoch<PK>>,
    ) -> Result<Self, GenerateError> {
        if host_height <= self.host_height {
            return Err(GenerateError::BadHostHeight);
        } else if host_timestamp <= self.host_timestamp {
            return Err(GenerateError::BadHostTimestamp);
        }

        let prev_block_hash = self.calc_hash();
        // If self defines a new epoch than the new block starts a new epoch
        // with epoch id equal to self’s block hash.  Otherwise, epoch doesn’t
        // change and the new block uses the same epoch id as self.
        let epoch_id = match self.next_epoch.is_some() {
            false => self.epoch_id.clone(),
            true => prev_block_hash.clone(),
        };
        Ok(Self {
            version: crate::common::VersionZero,
            prev_block_hash,
            block_height: self.block_height.next(),
            host_height,
            host_timestamp,
            state_root,
            epoch_id,
            next_epoch,
        })
    }

    /// Constructs a new genesis block.
    ///
    /// A genesis block is identified by previous block hash and epoch id both
    /// being all-zero hash.
    pub fn generate_genesis(
        block_height: crate::BlockHeight,
        host_height: crate::HostHeight,
        host_timestamp: u64,
        state_root: CryptoHash,
        next_epoch: crate::Epoch<PK>,
    ) -> Result<Self, GenerateError> {
        Ok(Self {
            version: crate::common::VersionZero,
            prev_block_hash: CryptoHash::DEFAULT,
            block_height,
            host_height,
            host_timestamp,
            state_root,
            epoch_id: CryptoHash::DEFAULT,
            next_epoch: Some(next_epoch),
        })
    }
}

#[test]
fn test_block_generation() {
    use crate::validators::{MockPubKey, MockSignature, MockSigner};

    // Generate a genesis block and test it’s behaviour.
    let genesis_hash = "Zq3s+b7x6R8tKV1iQtByAWqlDMXVVD9tSDOlmuLH7wI=";
    let genesis_hash = CryptoHash::from_base64(genesis_hash).unwrap();

    let genesis = Block::generate_genesis(
        crate::BlockHeight::from(0),
        crate::HostHeight::from(42),
        24,
        CryptoHash::test(66),
        crate::Epoch::test(&[(0, 10), (1, 10)]),
    )
    .unwrap();

    assert!(genesis.is_genesis());

    let mut block = genesis.clone();
    block.prev_block_hash = genesis_hash.clone();
    assert!(!block.is_genesis());

    let mut block = genesis.clone();
    block.epoch_id = genesis_hash.clone();
    assert!(!block.is_genesis());

    assert_eq!(genesis_hash, genesis.calc_hash());
    assert_ne!(genesis_hash, block.calc_hash());

    let pk = MockPubKey(77);
    let signer = MockSigner(pk);
    let signature = genesis.sign(&signer);
    assert_eq!(MockSignature(1722674425, pk), signature);
    assert!(genesis.verify(&pk, &signature));
    assert!(!genesis.verify(&MockPubKey(88), &signature));
    assert!(!genesis.verify(&pk, &MockSignature(0, pk)));

    let mut block = genesis.clone();
    block.host_timestamp += 1;
    assert_ne!(genesis_hash, block.calc_hash());
    assert!(!block.verify(&pk, &signature));

    // Try creating invalid next block.
    assert_eq!(
        Err(GenerateError::BadHostHeight),
        genesis.generate_next(
            crate::HostHeight::from(42),
            100,
            CryptoHash::test(99),
            None
        )
    );
    assert_eq!(
        Err(GenerateError::BadHostTimestamp),
        genesis.generate_next(
            crate::HostHeight::from(43),
            24,
            CryptoHash::test(99),
            None
        )
    );

    // Create next block and test its behaviour.
    let block = genesis
        .generate_next(
            crate::HostHeight::from(50),
            50,
            CryptoHash::test(99),
            None,
        )
        .unwrap();
    assert!(!block.is_genesis());
    assert_eq!(crate::BlockHeight::from(1), block.block_height);
    assert_eq!(genesis_hash, block.prev_block_hash);
    assert_eq!(genesis_hash, block.epoch_id);
    let hash = "uv7IaNMkac36VYAD/RNtDF14wY/DXxlxzsS2Qi+d4uw=";
    let hash = CryptoHash::from_base64(hash).unwrap();
    assert_eq!(hash, block.calc_hash());

    // Create next block within and introduce a new epoch.
    let epoch = Some(crate::Epoch::test(&[(0, 20), (1, 10)]));
    let block = block
        .generate_next(
            crate::HostHeight::from(60),
            60,
            CryptoHash::test(99),
            epoch,
        )
        .unwrap();
    assert_eq!(hash, block.prev_block_hash);
    assert_eq!(genesis_hash, block.epoch_id);
    let hash = "JWVBe5GotaDzyClzBuArPLjcAQTRElMCxvstyZ0bMtM=";
    let hash = CryptoHash::from_base64(hash).unwrap();
    assert_eq!(hash, block.calc_hash());

    // Create next block which belongs to the new epoch.
    let block = block
        .generate_next(
            crate::HostHeight::from(65),
            65,
            CryptoHash::test(99),
            None,
        )
        .unwrap();
    assert_eq!(hash, block.prev_block_hash);
    assert_eq!(hash, block.epoch_id);
}
