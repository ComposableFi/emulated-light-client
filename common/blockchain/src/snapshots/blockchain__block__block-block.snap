---
source: common/blockchain/src/block.rs
expression: check(&block)
---
[
    0,
    126,
    90,
    201,
    10,
    178,
    207,
    85,
    189,
    80,
    13,
    163,
    112,
    235,
    105,
    78,
    143,
    188,
    233,
    155,
    153,
    248,
    202,
    172,
    56,
    15,
    114,
    95,
    101,
    124,
    198,
    197,
    239,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    50,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    50,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    99,
    0,
    0,
    0,
    99,
    0,
    0,
    0,
    99,
    0,
    0,
    0,
    99,
    0,
    0,
    0,
    99,
    0,
    0,
    0,
    99,
    0,
    0,
    0,
    99,
    0,
    0,
    0,
    99,
    126,
    90,
    201,
    10,
    178,
    207,
    85,
    189,
    80,
    13,
    163,
    112,
    235,
    105,
    78,
    143,
    188,
    233,
    155,
    153,
    248,
    202,
    172,
    56,
    15,
    114,
    95,
    101,
    124,
    198,
    197,
    239,
    0,
]
